### PR TITLE
core(fr): extract storage and service worker driver methods

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
Splits out the storage and service worker bits from #12366 along with a couple of the super low value driver abstractions. Again the goal here is that FR-shared logic is refactored to accept a *session* instead of a driver. Ultimately everything left in `driver.js` and `gather-runner.js` should be specific to legacy flow with everything shared living elsewhere.

**Related Issues/PRs**
ref #11313 #12366

### CODE REVIEW

File: lighthouse-core/gather/driver/execution-context.js [hunk 1]
```diff
@@ -51,6 +52,9 @@ class ExecutionContext {
   async _getOrCreateIsolatedContextId() {
     if (typeof this._executionContextId === 'number') return this._executionContextId;
 
+    await this._session.sendCommand('Page.enable');
+    await this._session.sendCommand('Runtime.enable');
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver/execution-context.js:56]: before this dependency was an implicit global relationship, felt better to make it explicit.

File: lighthouse-core/gather/driver/service-workers.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+function getServiceWorkerRegistrations(session) {
+  return new Promise((resolve, reject) => {
+    session.once('ServiceWorker.workerRegistrationUpdated', data => {
+      session.sendCommand('ServiceWorker.disable').then(_ => resolve(data), reject);
+    });
+    session.sendCommand('ServiceWorker.enable').catch(reject);
+  });
+}
+
+module.exports = {getServiceWorkerVersions, getServiceWorkerRegistrations};
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver/service-workers.js:53]: nothing should be changed here

File: lighthouse-core/gather/driver/storage.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+  log.timeEnd(status);
+}
+
+module.exports = {
+  clearDataForOrigin,
+  cleanBrowserCaches,
+  getImportantStorageWarning,
+  UIStrings,
+};
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver/storage.js:121]: nothing should be different here either

File: lighthouse-core/gather/gather-runner.js [hunk 4]
```diff
@@ -119,6 +121,53 @@ class GatherRunner {
...
 
+  /**
+   * Rejects if any open tabs would share a service worker with the target URL.
+   * This includes the target tab, so navigation to something like about:blank
+   * should be done before calling.
+   * @param {LH.Gatherer.FRProtocolSession} session
+   * @param {string} pageUrl
+   * @return {Promise<void>}
+   */
+  static assertNoSameOriginServiceWorkerClients(session, pageUrl) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/gather-runner.js:132]: I moved this *here* instead of service-workers because it's going to be specific to legacy flow and not shared. We can't assert this in FR where we'll be *on* the page without any `about:blank` in between.

Given our neutering of the PWA checks I'm not sure it really has much value for us today anyway.

File: lighthouse-core/gather/gather-runner.js [hunk 5]
```diff
@@ -129,19 +178,38 @@ class GatherRunner {
     const status = {msg: 'Initializing…', id: 'lh:gather:setupDriver'};
     log.time(status);
     const resetStorage = !options.settings.disableStorageReset;
-    await driver.assertNoSameOriginServiceWorkerClients(options.requestedUrl);
+    const session = driver.defaultSession;
+
+    // Assert no service workers are still installed, so we test that they would actually be installed for a new user.
+    // TODO(FR-COMPAT): re-evaluate the necessity of this check
```
Comment (patrickhulce) [inline lighthouse-core/gather/gather-runner.js:184]: see above comment

File: lighthouse-core/gather/gather-runner.js [hunk 6]
```diff
@@ -129,19 +178,38 @@ class GatherRunner {
...
+    await GatherRunner.assertNoSameOriginServiceWorkerClients(session, options.requestedUrl);
+
+    // Emulate our target device screen and user agent.
     await emulation.emulate(driver.defaultSession, options.settings);
-    await driver.enableRuntimeEvents();
-    await driver.enableAsyncStacks();
+
+    // Enable `Debugger` domain to receive async stacktrace information on network request initiators.
+    // This is critical for tracing certain performance simulation situations.
+    session.on('Debugger.paused', () => session.sendCommand('Debugger.resume'));
```
Comment (patrickhulce) [inline lighthouse-core/gather/gather-runner.js:192]: this will be extracted along with the other shared driver setup in a future PR that focuses on gather-runner

File: lighthouse-core/test/gather/driver/service-workers-test.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+beforeEach(() => {
+  sessionMock = createMockSession();
+});
+
+afterEach(() => {
+  jest.useRealTimers();
+});
+
+describe('.getServiceWorkerVersions', () => {
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/driver/service-workers-test.js:24]: these weren't tested before 🎉

File: lighthouse-core/test/gather/gather-runner-test.js [hunk 8]
```diff
@@ -495,41 +596,6 @@ describe('GatherRunner', function() {
     expect(artifacts.devtoolsLogs).toHaveProperty('pageLoadError-nextPass');
   });
 
-  it('does not clear origin storage with flag --disable-storage-reset', () => {
```
Comment (patrickhulce) [inline lighthouse-core/test/gather/gather-runner-test.js:498]: this test was moved next to the "clears origin storage" test

**Comment by patrickhulce** [comment]
> wdyt of async/await'ing the FR/shared code?

Which methods did you have in mind? I agree the promise logic in service workers is hard to follow, but they're mostly event based which demand that structure (could maybe remove a `.catch(reject)` or two in exchange for a try/catch if that's what you mean?)

Storage methods I believe are all async'd up already.

**Comment by connorjclark** [comment]
oh, the only offender was `assertNoSameOriginServiceWorkerClients` which is in gather-runner. nvm.

File: lighthouse-core/test/gather/driver-test.js [hunk 9]
```diff
@@ -202,30 +202,6 @@ describe('.beginTrace', () => {
   });
 });
 
-describe('.setExtraHTTPHeaders', () => {
```
Comment (adamraine) [inline lighthouse-core/test/gather/driver-test.js:205]: Are these tests moved to `gather-runner-test.js`?

**Review by adamraine — APPROVED** [review]
Overall LGTM
